src/clj/opentracing_clj/core.clj

```diff
@@ -117,16 +117,22 @@
   ([^Span span ^String name]
    (.setOperationName span name)))
 
+(defn- resolve-tag-key ^String [key]
+  (cond (keyword? key) (name key)
+        (instance? Tag key) (.getKey ^Tag key)
+        :else ^String key))
+
 (defn set-tag
   "Sets a key/value tag on the Span."
-  ([^String key value]
+  ([key value]
    (with-active-span s
      (set-tag s key value)))
-  ([^Span span ^String key value]
+  ([^Span span key value]
    (cond
-     (instance? Boolean value) (.setTag span key ^Boolean value)
-     (instance? Number value)  (.setTag span key ^Number value)
-     :else                     (.setTag span key ^String (str value)))))
+     (instance? Boolean value) (.setTag span ^String (resolve-tag-key key) ^Boolean value)
+     (instance? Number value)  (.setTag span ^String (resolve-tag-key key) ^Number value)
+     (instance? Tag value)     (.setTag span ^String (resolve-tag-key key) ^String (.getKey ^Tag value))
```

Comment:
As I understand it, an io.opentracing.tag cannot be the value for another io.opentracing.tag.  The constants used for a tag value in the Tags object are all non-tags.